mino/minogrpc/server_test.go

```diff
@@ -1299,46 +1235,123 @@ func (t testMeshHandler) Stream(out mino.Sender, in mino.Receiver) error {
 	if !ok {
-		return xerrors.Errorf("failed to cast to dummy string: %v (type %T)", msg, msg)
+		return xerrors.Errorf("unexpeted first message: %T", msg)
 	}
 
-	var to mino.Address
-	// if I have no neighbor that means I am at the end of the ring. In that
-	// case I send the message to myself, so the message won't be relayed
-	// (because my own address is in the list of participants) and the receiver
-	// channel will be filled.
-	if t.neighborID == "" {
-		to = address{t.addrID}
-	} else {
-		to = address{t.neighborID}
+	if dummy.Value != "start" {
+		return xerrors.Errorf("expected start message, got '%s'", dummy.Value)
 	}
 
-	stringMsg := dummy.Value + "_" + t.addrID
+	for _, addr := range t.addrs {
 
-	dummyReturn := &wrappers.StringValue{Value: stringMsg}
-	errs := out.Send(dummyReturn, to)
+		rsleep()
+
+		if addr.String() == t.addr.String() {
+			continue
+		}
+
+		errs := out.Send(&wrappers.StringValue{Value: "first"}, addr)
+		err, more := <-errs
+		if more {
+			return xerrors.Errorf("unexpected error while sending first: %v", err)
+		}
+	}
+
+	firstAddrs := make([]string, len(t.addrs)-1)
+	for i := 0; i < len(t.addrs)-1; i++ {
+
+		rsleep()
+
+		from, msg, err := in.Recv(context.Background())
+		if err != nil {
+			return xerrors.Errorf("failed to receive first message: %v", err)
+		}
+
+		dummy, ok := msg.(*wrappers.StringValue)
+		if !ok {
+			return xerrors.Errorf("unexpected second message: %T", msg)
+		}
+
+		if dummy.Value != "first" {
+			return xerrors.Errorf("%s expected first message, got '%s'", t.addr, dummy.Value)
+		}
+		firstAddrs[i] = from.String()
+	}
+
+	// try to see if there is additional msg
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	from, msg, err := in.Recv(ctx)
+	if err == nil {
+		return xerrors.Errorf("%s got an aditional message from %s: %v", t.addr, from, msg)
+	}
+
+	sort.Strings(firstAddrs)
+	for i, j := 0, 0; i < len(firstAddrs); j++ {
+		rsleep()
+
+		if t.addr.String() == t.addrs[j].String() {
+			continue
+		}
+		if firstAddrs[i] != t.addrs[j].String() {
+			return xerrors.Errorf("expected '%s' to equal '%s'", firstAddrs[i], t.addrs[j].String())
+		}
+		i++
+	}
+
+	rsleep()
+
+	errs := out.Send(&wrappers.StringValue{Value: "finish"}, fromAddr)
 	err, more := <-errs
 	if more {
-		return xerrors.Errorf("got an error sending from the relay to the "+
-			"neighbor: %v", err)
+		return xerrors.Errorf("unexpected error while sending finsh: %v", err)
 	}
+
 	return nil
 }
+
+// Create server
+
+func createServer(t *testing.T, id string) (string, *address, *Server, Peer) {
+	identifier := id
+	addr := &address{
+		id: identifier,
+	}
+	server, err := NewServer(addr, TreeRoutingFactory)
+	require.NoError(t, err)
+	server.traffic.log = true
+	server.StartServer()
+	peer := Peer{
+		Address:     server.listener.Addr().String(),
+		Certificate: server.cert.Leaf,
+	}
+
+	return identifier, addr, server, peer
+}
+
+// Sleep random
+
+func rsleep() {
```

Comment:
I'm not a big fan of the random sleep. You expect from a unit test to make sure a particular scenario will end as expected. You manually send messages in different order to test that they will be handled as expected. Random sleep only provokes random errors on Travis.

Reply:
Yes I agree. I would however find interesting at some point to have some sort of fuzzing tests with different timing to stress the system. But of course not in this context.